### ISSUE TITLE
Saving VAE failed when using save_detector

### ISSUE THREAD

**Issue by mbisan** [SeldonIO/alibi-detect#695]
Hi, I'm trying to learn how to use the library, and I'm trying to train and save a VAE as seen in the [tutorial](https://docs.seldon.io/projects/alibi-detect/en/latest/examples/od_vae_cifar10.html), I'm just executing the code that builds the model and saves it after training.

This is the error I receive:

```
Traceback (most recent call last):
  File "/{}/.local/lib/python3.10/site-packages/alibi_detect/saving/saving.py", line 71, in save_detector
    save_detector_legacy(detector, filepath)
  File "/{}/.local/lib/python3.10/site-packages/alibi_detect/saving/tensorflow/_saving.py", line 237, in save_detector_legacy
    save_tf_vae(detector, filepath)
  File "/{}/.local/lib/python3.10/site-packages/alibi_detect/saving/tensorflow/_saving.py", line 786, in save_tf_vae
    detector.vae.encoder.encoder_net.save(model_dir.joinpath('encoder_net.h5'))
  File "/{}/.local/lib/python3.10/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/{}/.local/lib/python3.10/site-packages/keras/engine/base_layer.py", line 745, in get_config
    raise NotImplementedError(textwrap.dedent(f"""
NotImplementedError: 
Layer ModuleWrapper has arguments ['self', 'module', 'method_name']
in `__init__` and therefore must override `get_config()`.

Example:

class CustomLayer(keras.layers.Layer):
    def __init__(self, arg1, arg2):
        super().__init__()
        self.arg1 = arg1
        self.arg2 = arg2

    def get_config(self):
        config = super().get_config()
        config.update({
            "arg1": self.arg1,
            "arg2": self.arg2,
        })
        return config

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/{}/Documents/test_alibi_detect/test_alibi_script2.py", line 70, in <module>
    save_detector(od, filepath, legacy = True)
  File "/{}/.local/lib/python3.10/site-packages/alibi_detect/saving/saving.py", line 77, in save_detector
    raise RuntimeError(f'Saving failed. The save directory {filepath} has been cleaned.') from error
RuntimeError: Saving failed. The save directory outlierVAE has been cleaned.
```
The code I executed:

```
import os
import logging
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
tf.keras.backend.clear_session()
from tensorflow.python.keras.layers import Conv2D, Conv2DTranspose, Dense, Layer, Reshape, InputLayer
from tqdm import tqdm

from alibi_detect.models.tensorflow import elbo
from alibi_detect.od import OutlierVAE
from alibi_detect.utils.fetching import fetch_detector
from alibi_detect.saving import save_detector, load_detector

logger = tf.get_logger()
logger.setLevel(logging.ERROR)

train, test = tf.keras.datasets.cifar10.load_data()
X_train, y_train = train
X_test, y_test = test

X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255
print(X_train.shape, y_train.shape, X_test.shape, y_test.shape)

filepath = './'  # change to directory where model is downloaded
detector_type = 'outlier'
dataset = 'cifar10'
detector_name = 'outlierVAE'
filepath = os.path.join(filepath, detector_name)

latent_dim = 1024

encoder_net = tf.keras.Sequential(
  [
      InputLayer(input_shape=(32, 32, 3)),
      Conv2D(64, 4, strides=2, padding='same', activation=tf.nn.relu),
      Conv2D(128, 4, strides=2, padding='same', activation=tf.nn.relu),
      Conv2D(512, 4, strides=2, padding='same', activation=tf.nn.relu)
  ])

decoder_net = tf.keras.Sequential(
  [
      InputLayer(input_shape=(latent_dim,)),
      Dense(4*4*128),
      Reshape(target_shape=(4, 4, 128)),
      Conv2DTranspose(256, 4, strides=2, padding='same', activation=tf.nn.relu),
      Conv2DTranspose(64, 4, strides=2, padding='same', activation=tf.nn.relu),
      Conv2DTranspose(3, 4, strides=2, padding='same', activation='sigmoid')
  ])

# initialize outlier detector
od = OutlierVAE(threshold=.015,  # threshold for outlier score
                score_type='mse',  # use MSE of reconstruction error for outlier detection
                encoder_net=encoder_net,  # can also pass VAE model instead
                decoder_net=decoder_net,  # of separate encoder and decoder
                latent_dim=latent_dim,
                samples=2)
# train
od.fit(X_train[:100],
        loss_fn=elbo,
        cov_elbo=dict(sim=.05),
        epochs=1,
        verbose=True)

# save the trained outlier detector
save_detector(od, filepath, legacy = True)
```

The installation is fresh via PIP,  `pip install alibi-detect[tensorflow]`

**Comment by mbisan** [comment]
Hi, just tried downgrading tensorflow to 2.6.2 and now the functionality works fine.

**Comment by ascillitoe** [comment]
Hi @mbisan. Thank you for reporting this, I am going to reopen this until we can get to bottom of why the error was occurring when you had a newer `tensorflow` version installed.

**Comment by ascillitoe** [comment]
Hi again @mbisan. It looks like the error you experienced might be occurring because private versions of the Keras classes (`Dense`, `Layer` etc) are being used to create the `encoder_net` and `decoder_net` models. In other words, the line:

```python
from tensorflow.python.keras.layers import Conv2D, Conv2DTranspose, Dense, Layer, Reshape, InputLayer
```

should be (see [this comment](https://github.com/tensorflow/tensorflow/issues/33075#issuecomment-539070546)):

```python
from tensorflow.keras.layers import Conv2D, Conv2DTranspose, Dense, Layer, Reshape, InputLayer
```

Hopefully with the latter, you should be able to use newer `tensorflow` versions again. 

Also, just to check, regarding the `from tensorflow.python...`, did you make this change because you were experiencing other errors? Or perhaps because your IDE was throwing a warning? There is a [known issue](https://youtrack.jetbrains.com/issue/PY-53599/tensorflow.keras-subpackages-are-unresolved) with editors such as PyCharm incorrectly throwing a warning when the correct `from tensorflow.keras` is used.

**Comment by ascillitoe** [comment]
p.s. @mbisan I will close this issue again now, but it would be great if you could reopen it if you continue to have problems when the latter import command is used.

**Comment by mbisan** [comment]
Hi @ascillitoe I was wrong about how I fixed the problem, I downgraded tensorflow AND fixed the inport as you did (but didn't realize about the import).

I've been trying to make this work in two different computers and I copied the example again, and there the import is indeed `tensorflow.keras` instead of `tensorflow.python.keras`. I went back to the other computer and couldn't get it to work, the import is incorrect.

Thank you!
